# Spam Detector

In [4]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [5]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

My prediction (educated guess) on which model will perform better between Logistical Regression and Random Forest Classifier, is the Random Forest Classifier.  I weigh the Random Forest more because it is based on the decision tree 
method, and breaks the decision trees into many smaller trees which in and of itself, stregthens it against weak learners and overfitting.  It is trained on many different pieces of the data, where Logical Regression is trained on a limit split of the data.  


## Split the Data into Training and Testing Sets

In [6]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"].ravel()   # set y to nd.array representing the spam column of the dataframe named data using the ravel function 


X = data.copy()         # copy the data DataFrame to X
X.drop("spam", axis=1, inplace=True)    #  Drop the spam column from the copied dataframe 'X' leaving the remaining feature columns


In [7]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
data['spam'].value_counts()    # tally the values in the spam column of the data DataFrame

spam
0    2788
1    1813
Name: count, dtype: int64

In [8]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [9]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()    # scaler is now an instance of the StandardScaler


In [10]:
# Fit the Standard Scaler with the training data
X_train_fitted  = scaler.fit(X_train)    #  X_train_fitted is now a StandardScaler with the split X_train data fitted to it
X_test_fitted  = scaler.fit(X_test)    #  X_test_fitted is now a StandardScaler with the split X_test data fitted to it

In [11]:
# Scale the training data
X_train_scaled = X_train_fitted.transform(X_train)   # X_train_scaled contains the scaled(transformed) X_train portion of the split data 
X_test_scaled = X_test_fitted.transform(X_test)      # X_test_scaled contains the scaled(transformed) X_test portion of the split data

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [12]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
logistic_regression_model = LogisticRegression(random_state=1)           # Instantiate the logistic regression model (logistic_regression_model)
logistic_regression_model


LogisticRegression(random_state=1)

In [13]:
# Make and save testing predictions with the saved logistic regression model using the test data
logistic_regression_model.fit(X_train_scaled, y_train)                       # Fit the Random Forest model with the X y training data


testing_predictions = logistic_regression_model.predict(X_test_scaled)       # make the prediction against test data using the Logistic Regression  model
 


# Review the predictions
testing_predictions

array([0, 0, 1, ..., 0, 0, 0], dtype=int64)

In [14]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)             # Calculate the score via the accuracy_score function on the predicted outcome from the Logistic Regression  model

0.9313640312771503

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [15]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

rmf = RandomForestClassifier(random_state=1, n_estimators=500)  #  Instantiate the Random Forest Classifier model (rmf)
rmf


RandomForestClassifier(n_estimators=500, random_state=1)

In [16]:
# Make and save testing predictions with the saved logistic regression model using the test data

rmf.fit(X_train_scaled, y_train)                          # Fit the Random Forest model with the X y training data

testing_predictions = rmf.predict(X_test_scaled)          # make the prediction using the same test data using the Random Forest model

# Review the predictions
testing_predictions                 

array([0, 0, 1, ..., 0, 0, 1], dtype=int64)

In [17]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predictions)  #Calculate the score via the accuracy_score function on the predicted outcome from the Random Forest model

0.9617723718505647

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

The Random Forest model was the better performer.  Accuracy scores......
                                                                                  Random Forest model = 0.9617723718505647
                                                                          Logistical Regression model = 0.9313640312771503
                                                                    Random Forest performed better by   0.0304083405734144  (subject to vary some on each run due to random setting)

This confirmed my prediction that Random Forest would perform better and according to this past weeks lecture and course material, 
   these are the reasons that back it up as being better......  

   1 "Robust against overfitting as all weak learners are trained on different pieces of the data"
   2 "Can rank the importance of input variables in a natural way"
   3 "Can handle thousands of input variables without variable deletion"
   4 "Robust to outliers and nonlinear data"
   5 "Run efficiently on large datasets"

One of the main reasons that the Random Forest model accomplishes this, is due to the break down of decision trees into smaller decision trees.  The data in these smaller trees is a subset of the larger and makes decision making a much simpler and more accurate effort.  (This btw, is great to know!)                                          